docs/design/storage-auto-scale.md

```diff
@@ -0,0 +1,94 @@
+# Storage Automatic Scaler
+
+# Summary
+
+This feature would help users to get rid of maintenance task of keeping the system with enough available free space.
+In virtualized environments, we can scale the storage devices OSDs so this feature helps in automatic scaling. 
+
+# Proposal
+
+Implement auto storage scale CR per device class of a specific Storagecluster.
+
+# High-Level Design
+
+- Create a StorageAutoScale CR per Storagecluster device class
+    ```
+    apiVersion: ocs.openshift.io/v1
+    Kind: StorageAutoScaling
+    Metadata:
+        name: <storageclusterName>-<deviceClassName>
+        namespace: openshift-storage
+    Spec:
+        storageClusterName: <storageclusterName>
+        deviceClassName: <deviceClassName>
+        capacityLimit: <capacity>   
+    Status:
+        phase: (waiting/inProgress/completed/failure)
+        LastRunTimeStamp: (only on completed/failure)
+        currentOSDSize: 2TiB
+        expectedOSDSize: 4TiB
+        currentOSDCount: 3​
+        expectedOSDCount: 3
+    ```
+
+- Watch prometheus periodically after 10 mins and depending on the OSD fill percentage, reconcile the StorageAutoScaling CR.
+  
+- If the OSD size is less than 8Tib, do vertical scaling by doubling the each osd sizes for that device class.
+
+- If the OSD size is equal to 8Tib, do a horizontal scaling, by adding 1 osd of 8Tib on each storage node, by simply increasing the (count+1) in the device set associated with a device class.   
+
+# Low-Level Design 
+
+1) Add a single go-routine to watch periodically(10mins) all the OSDs sizes using Prometheus metrics in the storagecluster namespace.
+   1) With every run calculate different device class highest osd filled percentage.
+       `OsdPercentage`:
+        ```
+        ssd               69%
+        nvme               71%
+        ...
+        ```
+    2) If the highest osd percentage reported is more than 73-74%(reaching osd nearfull), scaling is needed.
+    3) If scaling is needed calculate the `expectedOSDSize` and `expectedOSDCount` per device class.
+       1)  Calculate the total number osds and osd size per device class from the prometheus response.
+       2)  Calculate the total number osds and osd size per device class from the storageCluster.
+       3)  If the above values are equal(needed to avoid double re-sizing), create or update the sync-map mapping:
```

Comment:
`Go routine would update the sync-map` is the step below. Isn't this redundant?

Reply:
correct, merge it